File: lib/features/search/data/user_search_repository.dart
//lib/features/search/data/user_search_repository.dart
import 'dart:async';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';

import '../../../core/services/country/country_resolver_service.dart';
import '../models/user_search_entry.dart';

class UserSearchRepository {
  UserSearchRepository({FirebaseFirestore? firestore, FirebaseAuth? auth})
      : _firestore = firestore ?? FirebaseFirestore.instance,
        _auth = auth ?? FirebaseAuth.instance;

  final FirebaseFirestore _firestore;
  final FirebaseAuth _auth;

  CollectionReference<Map<String, dynamic>> get _col =>
      _firestore.collection('user_search');

  Future<void> syncSelfIndex({
    required String displayName,
    required String shareId,
    required String game,
    required String badge,
    required String avatarUrl,
    String? country,
    String? usernameLower,
  }) async {
    final uid = _auth.currentUser?.uid.trim() ?? '';
    if (uid.isEmpty) return;

    try {
      final payload = <String, dynamic>{
        'userId': uid,
        'displayName': displayName.trim(),
        'displayNameLower': displayName.trim().toLowerCase(),
        'shareId': shareId.trim(),
        'shareIdLower': shareId.trim().toLowerCase(),
        'game': game.trim(),
        'badge': badge.trim(),
        'avatarUrl': avatarUrl.trim(),
        'updatedAtMs': DateTime.now().millisecondsSinceEpoch,
      };

      final trimmedCountry = (country ?? '').trim().toUpperCase();
      if (trimmedCountry.isNotEmpty) {
        payload['country'] = trimmedCountry;
      }

      final trimmedUsername = (usernameLower ?? '').trim().toLowerCase();
      if (trimmedUsername.isNotEmpty) {
        payload['usernameLower'] = trimmedUsername;
      }

      await _col.doc(uid).set(payload, SetOptions(merge: true)).timeout(const Duration(seconds: 12));
    } catch (e) {
      if (kDebugMode) {
        debugPrint('[UserSearchRepository] syncSelfIndex failed (non-fatal): $e');
      }
    }
  }

  Future<void> resyncSelfFromSources({
    required String displayName,
    required String shareId,
    required String avatarUrl,
    required String game,
    required String badge,
    String? country,
    String? usernameLower,
  }) => syncSelfIndex(
        displayName: displayName,
        shareId: shareId,
        game: game,
        badge: badge,
        avatarUrl: avatarUrl,
        country: country,
        usernameLower: usernameLower,
      );

  /// Best-effort, called immediately after UserProfileRepository commits
  /// a username reservation (both manual edit and auto-assignment), so
  /// the search index stays in sync without every caller having to know
  /// about it. Writes ONLY userId/usernameLower/updatedAtMs — never
  /// throws.
  Future<void> syncUsername(String usernameLower) async {
    final uid = _auth.currentUser?.uid.trim() ?? '';
    final lower = usernameLower.trim().toLowerCase();
    if (uid.isEmpty || lower.isEmpty) return;

    try {
      await _col.doc(uid).set(
        <String, dynamic>{
          'userId': uid,
          'usernameLower': lower,
          'updatedAtMs': DateTime.now().millisecondsSinceEpoch,
        },
        SetOptions(merge: true),
      ).timeout(const Duration(seconds: 12));
    } catch (e) {
      if (kDebugMode) {
        debugPrint('[UserSearchRepository] syncUsername failed (non-fatal): $e');
      }
    }
  }

  Future<void> backfillCountryIfMissing() async {
    final uid = _auth.currentUser?.uid.trim() ?? '';
    if (uid.isEmpty) return;

    try {
      final doc = await _col
          .doc(uid)
          .get(const GetOptions(source: Source.server))
          .timeout(const Duration(seconds: 10));

      final existing = (doc.data()?['country'] as String? ?? '').trim();
      if (existing.isNotEmpty) return;

      final resolved = await CountryResolverService.instance.resolveCountryCode();
      final trimmed = resolved.trim().toUpperCase();
      if (trimmed.isEmpty) return;

      await _col.doc(uid).set(
        <String, dynamic>{
          'userId': uid,
          'country': trimmed,
          'updatedAtMs': DateTime.now().millisecondsSinceEpoch,
        },
        SetOptions(merge: true),
      ).timeout(const Duration(seconds: 12));

      if (kDebugMode) {
        debugPrint('[UserSearchRepository] backfillCountryIfMissing: set country=$trimmed for $uid');
      }
    } catch (e) {
      if (kDebugMode) {
        debugPrint('[UserSearchRepository] backfillCountryIfMissing failed (non-fatal): $e');
      }
    }
  }

  Future<List<UserSearchEntry>> fetchNearby({
    required String countryCode,
    int limit = 20,
  }) async {
    final cc = countryCode.trim().toUpperCase();
    if (cc.isEmpty) return const [];

    final selfUid = _auth.currentUser?.uid.trim() ?? '';

    try {
      final snap = await _col
          .where('country', isEqualTo: cc)
          .orderBy('updatedAtMs', descending: true)
          .limit(limit + 1)
          .get(const GetOptions(source: Source.server))
          .timeout(const Duration(seconds: 15));

      final out = <UserSearchEntry>[];
      for (final d in snap.docs) {
        if (d.id == selfUid) continue;
        out.add(UserSearchEntry.fromDoc(d));
        if (out.length >= limit) break;
      }
      return out;
    } catch (e) {
      if (kDebugMode) {
        debugPrint('[UserSearchRepository] fetchNearby failed: $e');
      }
      return const [];
    }
  }

  Future<List<UserSearchEntry>> search(String query, {int limit = 20}) async {
    final trimmed = query.trim();
    if (trimmed.isEmpty) return const [];

    var lower = trimmed.toLowerCase();
    if (lower.startsWith('@')) lower = lower.substring(1);

    try {
      final results = <String, UserSearchEntry>{};

      final nameSnap = await _col
          .orderBy('displayNameLower')
          .startAt([lower])
          .endAt(['$lower\uf8ff'])
          .limit(limit)
          .get(const GetOptions(source: Source.server))
          .timeout(const Duration(seconds: 15));

      for (final d in nameSnap.docs) {
        final entry = UserSearchEntry.fromDoc(d);
        results[entry.userId] = entry;
      }

      // Exact username match — trigger for @-prefixed queries or once
      // enough characters have been typed that it's plausibly a username.
      if (trimmed.startsWith('@') || lower.length >= 3) {
        try {
          final usernameSnap = await _col
              .where('usernameLower', isEqualTo: lower)
              .limit(5)
              .get(const GetOptions(source: Source.server))
              .timeout(const Duration(seconds: 12));

          for (final d in usernameSnap.docs) {
            final entry = UserSearchEntry.fromDoc(d);
            results[entry.userId] = entry;
          }
        } catch (e) {
          if (kDebugMode) {
            debugPrint('[UserSearchRepository] username search failed: $e');
          }
        }
      }

      if (trimmed.length >= 3) {
        final idSnap = await _col
            .where('shareIdLower', isEqualTo: lower)
            .limit(5)
            .get(const GetOptions(source: Source.server))
            .timeout(const Duration(seconds: 12));

        for (final d in idSnap.docs) {
          final entry = UserSearchEntry.fromDoc(d);
          results[entry.userId] = entry;
        }

        if (results.isEmpty) {
          final idSnapExact = await _col
              .where('shareId', isEqualTo: trimmed)
              .limit(5)
              .get(const GetOptions(source: Source.server))
              .timeout(const Duration(seconds: 12));

          for (final d in idSnapExact.docs) {
            final entry = UserSearchEntry.fromDoc(d);
            results[entry.userId] = entry;
          }
        }
      }

      return results.values.toList(growable: false);
    } catch (e) {
      if (kDebugMode) {
        debugPrint('[UserSearchRepository] search failed: $e');
      }
      return const [];
    }
  }
}

File: web_client/lib/search/userSearchRepository.ts
import { collection, getDocs, getDoc, query, where, orderBy, limit, startAt, endAt, doc, setDoc } from 'firebase/firestore';
import { db, auth } from '@/lib/firebase';
import { resolveCountryCodeWeb } from '@/lib/countryResolver';

export interface UserSearchEntry {
  userId: string;
  displayName: string;
  shareId: string;
  game: string;
  badge: string;
  avatarUrl: string;
  country: string;
  usernameLower: string;
}

export async function fetchNearbyTeamsWeb(countryCode: string, selfUid: string, limitCount = 20): Promise<UserSearchEntry[]> {
  const cc = countryCode.trim().toUpperCase();
  if (!cc) return [];

  try {
    const q = query(
      collection(db, 'user_search'),
      where('country', '==', cc),
      orderBy('updatedAtMs', 'desc'),
      limit(limitCount + 1)
    );
    
    const snap = await getDocs(q);
    const results: UserSearchEntry[] = [];
    
    snap.forEach(d => {
      if (d.id !== selfUid) {
        results.push({ userId: d.id, ...d.data() } as UserSearchEntry);
      }
    });

    return results.slice(0, limitCount);
  } catch (err) {
    console.error('[UserSearchRepository] fetchNearbyTeamsWeb failed:', err);
    return [];
  }
}

export async function searchUsersWeb(searchQuery: string, selfUid: string, limitCount = 20): Promise<UserSearchEntry[]> {
  const trimmed = searchQuery.trim();
  let lower = trimmed.toLowerCase();
  if (lower.startsWith('@')) lower = lower.substring(1);
  
  if (!lower) return [];

  try {
    const resultsMap = new Map<string, UserSearchEntry>();

    // 1. Prefix match on Display Name
    const nameQ = query(
      collection(db, 'user_search'),
      orderBy('displayNameLower'),
      startAt(lower),
      endAt(lower + '\uf8ff'),
      limit(limitCount)
    );
    const nameSnap = await getDocs(nameQ);
    nameSnap.forEach(d => resultsMap.set(d.id, { userId: d.id, ...d.data() } as UserSearchEntry));

    // 2. Exact match on Username (triggers for @ prefix or >= 3 chars)
    if (trimmed.startsWith('@') || lower.length >= 3) {
      const usernameQ = query(
        collection(db, 'user_search'),
        where('usernameLower', '==', lower),
        limit(5)
      );
      const usernameSnap = await getDocs(usernameQ);
      usernameSnap.forEach(d => resultsMap.set(d.id, { userId: d.id, ...d.data() } as UserSearchEntry));
    }

    // 3. Exact match on Share ID
    if (trimmed.length >= 3) {
      const idQ = query(
        collection(db, 'user_search'),
        where('shareIdLower', '==', lower),
        limit(5)
      );
      const idSnap = await getDocs(idQ);
      idSnap.forEach(d => resultsMap.set(d.id, { userId: d.id, ...d.data() } as UserSearchEntry));
      
      // Fallback exact match on raw shareId if needed
      if (resultsMap.size === 0) {
        const idExactQ = query(
          collection(db, 'user_search'),
          where('shareId', '==', trimmed),
          limit(5)
        );
        const idExactSnap = await getDocs(idExactQ);
        idExactSnap.forEach(d => resultsMap.set(d.id, { userId: d.id, ...d.data() } as UserSearchEntry));
      }
    }

    // Filter out self and convert to array
    const finalResults = Array.from(resultsMap.values()).filter(u => u.userId !== selfUid);
    return finalResults.slice(0, limitCount);
  } catch (err) {
    console.error('[UserSearchRepository] searchUsersWeb failed:', err);
    return [];
  }
}

/**
 * Mirrors UserSearchRepository.syncSelfIndex() — keeps this user's
 * `user_search/{uid}` doc (the index every search/nearby query above
 * reads) up to date. Called whenever the fields it covers change:
 * display name, avatar, or (indirectly, via the profile save flow on
 * mobile) game. Best-effort — never throws, matching the Dart repo's
 * "non-fatal" background-sync behavior.
 */
export async function syncSelfIndexWeb(params: {
  displayName: string;
  shareId: string;
  game?: string;
  badge?: string;
  avatarUrl: string;
  country?: string;
  usernameLower?: string;
}) {
  const uid = auth.currentUser?.uid.trim();
  if (!uid) return;

  try {
    const displayName = params.displayName.trim() || (auth.currentUser?.displayName ?? '').trim();
    const shareId = params.shareId.trim();

    const payload: Record<string, unknown> = {
      userId: uid,
      displayName,
      displayNameLower: displayName.toLowerCase(),
      shareId,
      shareIdLower: shareId.toLowerCase(),
      game: (params.game ?? '').trim(),
      badge: (params.badge ?? '').trim(),
      avatarUrl: params.avatarUrl.trim(),
      updatedAtMs: Date.now(),
    };

    const country = (params.country ?? '').trim().toUpperCase();
    if (country) payload.country = country;

    const usernameLower = (params.usernameLower ?? '').trim().toLowerCase();
    if (usernameLower) payload.usernameLower = usernameLower;

    await setDoc(doc(db, 'user_search', uid), payload, { merge: true });
  } catch (err) {
    console.warn('[userSearchRepository] syncSelfIndexWeb failed (non-fatal):', err);
  }
}

/**
 * Mirrors UserSearchRepository.syncUsername() — writes ONLY
 * userId/usernameLower/updatedAtMs, called right after a username
 * reservation commits (both an explicit edit and lazy auto-assignment)
 * so the search index never drifts from the authoritative username doc.
 */
export async function syncUsernameWeb(usernameLower: string) {
  const uid = auth.currentUser?.uid.trim();
  const lower = usernameLower.trim().toLowerCase();
  if (!uid || !lower) return;

  try {
    await setDoc(
      doc(db, 'user_search', uid),
      { userId: uid, usernameLower: lower, updatedAtMs: Date.now() },
      { merge: true },
    );
  } catch (err) {
    console.warn('[userSearchRepository] syncUsernameWeb failed (non-fatal):', err);
  }
}

/**
 * Mirrors UserSearchRepository.backfillDisplayNameIfMissing() — a
 * self-heal for accounts whose user_search doc still has a blank name
 * (e.g. it was never synced before this fix existed). Only ever touches
 * displayName/displayNameLower, so it can never clobber other fields.
 */
export async function backfillDisplayNameIfMissingWeb(fallbackDisplayName: string) {
  const uid = auth.currentUser?.uid.trim();
  const name = fallbackDisplayName.trim();
  if (!uid || !name) return;

  try {
    const snap = await getDoc(doc(db, 'user_search', uid));
    const existing = typeof snap.data()?.displayName === 'string' ? snap.data()!.displayName.trim() : '';
    if (existing) return;

    await setDoc(
      doc(db, 'user_search', uid),
      { userId: uid, displayName: name, displayNameLower: name.toLowerCase(), updatedAtMs: Date.now() },
      { merge: true },
    );
  } catch (err) {
    console.warn('[userSearchRepository] backfillDisplayNameIfMissingWeb failed (non-fatal):', err);
  }
}

/**
 * Mirrors UserSearchRepository.backfillCountryIfMissing() — self-heals
 * "Teams Near You" eligibility for accounts that predate the country
 * field, or never triggered syncSelfIndexWeb.
 */
export async function backfillCountryIfMissingWeb() {
  const uid = auth.currentUser?.uid.trim();
  if (!uid) return;

  try {
    const snap = await getDoc(doc(db, 'user_search', uid));
    const existing = typeof snap.data()?.country === 'string' ? snap.data()!.country.trim() : '';
    if (existing) return;

    const resolved = (await resolveCountryCodeWeb()).trim().toUpperCase();
    if (!resolved) return;

    await setDoc(
      doc(db, 'user_search', uid),
      { userId: uid, country: resolved, updatedAtMs: Date.now() },
      { merge: true },
    );
  } catch (err) {
    console.warn('[userSearchRepository] backfillCountryIfMissingWeb failed (non-fatal):', err);
  }
}
